### PR TITLE
Add end-to-end test framework inspired from iOS interactive tests

### PR DESCRIPTION
**Pull request by Mindfulplays**
This brings interactive tests from iOS and allows them to be run in any environment. 
Specifically, given a set of `Game` classes under `Tests/Interactive/Tests`, we can run that `Game` instance in a test harness under DesktopGL (Mac/Windows)/iOS (Android coming next).

- Lots of reshuffling from Tests/Interactive/iOS -> Common/Tests
  - Also added doc to run the tests in simulator/on-device.
- New TestRunners (for now DesktopGL on Mac/Windows and iOS)- will add Android next
  - DesktopGL only allows running one test per run, due to STA/SDL Window event loop.
  - Perhaps consider NUnit test that launches the DesktopGL in a separate process with the correct command-line parameter?
- Added doc in Tests/Interactive/README.md on how to run the test runners.
- Allows 'automatic' running of InteractiveTests too (i.e. automatically run one test, allow test to signal 'finish' and exit)
  - i.e. provided a `TestGame` that implements `Game` and allows a test author to signal when the test is done. (i.e. either by clicking the Exit button or the test can run automatically, then signal Exit)

### CODE REVIEW

**Comment by Mindfulplays** [comment]
@tomspilman @harry-cpp Please take a look

**Comment by Mindfulplays** [comment]
Squashed commits and cleaned up comments/code etc. Tested on DesktopGL (mac/windows) + iOS

**Comment by Mindfulplays** [comment]
This PR is now ready for you folks, please take a look.

I have more coming up that adds RenderTarget/etc tests to this and also adds WindowsDX/Android runners etc. it's been very useful to understand/verify behaviors as I have made Metal work using these tests in parallel.

Not sure how I magically summon reviewers but here we go: @SimonDarksideJ @harry-cpp @tomspilman :)
Thanks!

**Comment by Mindfulplays** [comment]
Could someone take a look? @mrhelmut

**Comment by Mindfulplays** [comment]
I can resolve conflicts (mostly license/doc?) but could someone please take a look? This was super useful in testing out/verifying Metal, and hence I think for the proposed GDK/Vulkan/DX12 stuff it will be beneficial as well. @mrhelmut @harry-cpp

**Comment by Mindfulplays** [comment]
@SimonDarksideJ FYI...

**Comment by SimonDarksideJ** [comment]
Looks interesting @Mindfulplays , can you resolve the conflicts so we can get a report from the build server environment and then we can compare.
It does look very interesting.

File: Tests/Interactive/Common/GameDebug.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    public partial class GameDebug
+    {
+        /// <summary>Output a single console message.</summary>
+        public static void C(string message)
+        {
+            System.Console.WriteLine($"MGDBG: {message}");
+        }
+
+        /// <summary>Output an error message to the console.</summary>
+        public static void E(string message)
```
Comment (dellis1972) [inline Tests/Interactive/Common/GameDebug.cs:25]: Could we get more descriptive method names for these? `LogError` or `LogDebug`

**Comment by Mindfulplays** [comment]
This PR is too out-of-date, so moved to https://github.com/MonoGame/MonoGame/pull/8390

File: Tests/Interactive/Common/GameDebug.cs (hunk 1 above)
Comment (Mindfulplays) [reply Tests/Interactive/Common/GameDebug.cs]: I am mimicking Log.e/i from Android where it's easy to write/ready in a concise manner.
In particular, I am using these as `GD.E` or `GD.I` and hoping the xmldoc helps clarify...

**Pull request closed without merging**